{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div style=\"font-weight: bold; color:#5D8AA8\" align=\"center\">\n",
    "    <div style=\"font-size: xx-large; color:gray\"> Annotation Results analysis </div><br> </center>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from IPython import display\n",
    "import os\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Remember working directory: /home/ines/Escritorio/epigCTCF\n"
     ]
    }
   ],
   "source": [
    "# RUN THIS NOTEBOOK FROM THE Modelo_11_estados FOLDER!\n",
    "print(f\"Remember working directory: {os.getcwd()}\")\n",
    "input_folder=\"my_results/GO_enrichment\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### (1) **GO enrichment files**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### GO: **Biological Process**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "biological_processes_df.shape: (318, 6)\n",
      "biological_processes_df.columns: ['Term', 'Annotated', 'Significant', 'Expected', 'Fisher Test', 'Significant Genes']\n"
     ]
    }
   ],
   "source": [
    "biological_processes_df = pd.read_csv(f\"{input_folder}/BP_enriched_genes_entrez.txt.tsv\", sep=\"\\t\", index_col=0, header=\"infer\")\n",
    "print(f\"biological_processes_df.shape: {biological_processes_df.shape}\")\n",
    "print(f\"biological_processes_df.columns: {list(biological_processes_df.columns)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of unique_terms: 313\n"
     ]
    }
   ],
   "source": [
    "unique_terms_CB = pd.unique(biological_processes_df[\"Term\"])\n",
    "print(f\"Number of unique_terms: {len(unique_terms_CB)}\")\n",
    "# for unique_term in unique_terms_CB: print(unique_term)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\t* plasma membrane bounded cell projection ...\n",
      "\t* cell projection organization\n",
      "\t* cellular component organization\n",
      "\t* cell development\n",
      "\t* cellular component organization or bioge...\n",
      "\t* localization\n",
      "\t* cellular component morphogenesis\n",
      "\t* nervous system development\n",
      "\t* cell morphogenesis\n",
      "\t* cytoskeleton organization\n",
      "\t* anatomical structure development\n",
      "\t* actin filament-based process\n",
      "\t* movement of cell or subcellular componen...\n",
      "\t* multicellular organism development\n",
      "\t* generation of neurons\n",
      "\t* cell projection assembly\n",
      "\t* neurogenesis\n",
      "\t* developmental process\n",
      "\t* actin cytoskeleton organization\n",
      "\t* neuron projection development\n",
      "\t* neuron development\n",
      "\t* cell adhesion\n",
      "\t* regulation of signaling\n",
      "\t* biological adhesion\n",
      "\t* neuron differentiation\n",
      "\t* system development\n",
      "\t* regulation of cell communication\n",
      "\t* neuron projection morphogenesis\n",
      "\t* regulation of small GTPase mediated sign...\n",
      "\t* cell projection morphogenesis\n",
      "\t* supramolecular fiber organization\n",
      "\t* cell part morphogenesis\n",
      "\t* regulation of cellular component organiz...\n",
      "\t* regulation of localization\n",
      "\t* ion transport\n",
      "\t* ion transmembrane transport\n",
      "\t* anatomical structure morphogenesis\n",
      "\t* microtubule-based movement\n",
      "\t* multicellular organismal process\n",
      "\t* cellular developmental process\n",
      "\t* organelle organization\n",
      "\t* intracellular signal transduction\n",
      "\t* homophilic cell adhesion via plasma memb...\n",
      "\t* cell morphogenesis involved in different...\n",
      "\t* modulation of chemical synaptic transmis...\n",
      "\t* regulation of trans-synaptic signaling\n",
      "\t* cation transmembrane transport\n",
      "\t* regulation of biological quality\n",
      "\t* positive regulation of GTPase activity\n",
      "\t* regulation of plasma membrane bounded ce...\n",
      "\t* regulation of GTPase activity\n",
      "\t* establishment of localization\n",
      "\t* small GTPase mediated signal transductio...\n",
      "\t* regulation of cell projection organizati...\n",
      "\t* cell morphogenesis involved in neuron di...\n",
      "\t* cellular component assembly\n",
      "\t* microtubule-based process\n",
      "\t* cation transport\n",
      "\t* inorganic ion transmembrane transport\n",
      "\t* cell differentiation\n",
      "\t* cilium organization\n",
      "\t* locomotion\n",
      "\t* transmembrane transport\n",
      "\t* metal ion transport\n",
      "\t* cell junction organization\n",
      "\t* cilium assembly\n",
      "\t* transport\n",
      "\t* regulation of signal transduction\n",
      "\t* regulation of multicellular organismal p...\n",
      "\t* regulation of Ras protein signal transdu...\n",
      "\t* regulation of cell development\n",
      "\t* actin filament organization\n",
      "\t* cilium movement\n",
      "\t* positive regulation of catalytic activit...\n",
      "\t* inorganic cation transmembrane transport\n",
      "\t* regulation of nervous system development\n",
      "\t* synaptic signaling\n",
      "\t* cell motility\n",
      "\t* localization of cell\n",
      "\t* regulation of neurogenesis\n",
      "\t* chemical synaptic transmission\n",
      "\t* anterograde trans-synaptic signaling\n",
      "\t* trans-synaptic signaling\n",
      "\t* cell-cell signaling\n",
      "\t* positive regulation of molecular functio...\n",
      "\t* regulation of transmembrane transport\n",
      "\t* organelle assembly\n",
      "\t* regulation of intracellular signal trans...\n",
      "\t* cell-cell adhesion via plasma-membrane a...\n",
      "\t* regulation of membrane potential\n",
      "\t* cell-cell adhesion\n",
      "\t* regulation of cellular component biogene...\n",
      "\t* regulation of transport\n",
      "\t* Ras protein signal transduction\n",
      "\t* adherens junction organization\n",
      "\t* sodium ion transmembrane transport\n",
      "\t* extracellular matrix organization\n",
      "\t* activation of GTPase activity\n",
      "\t* cellular component biogenesis\n",
      "\t* photoreceptor cell maintenance\n",
      "\t* actin filament bundle assembly\n",
      "\t* regulation of neuron differentiation\n",
      "\t* actomyosin structure organization\n",
      "\t* actin filament bundle organization\n",
      "\t* regulation of neuron projection developm...\n",
      "\t* developmental growth\n",
      "\t* sodium ion transport\n",
      "\t* regulation of ion transmembrane transpor...\n",
      "\t* positive regulation of hydrolase activit...\n",
      "\t* regulation of cellular component movemen...\n",
      "\t* cell-substrate adhesion\n",
      "\t* regulation of transporter activity\n",
      "\t* cell-matrix adhesion\n",
      "\t* axonogenesis\n",
      "\t* macromolecule localization\n",
      "\t* central nervous system development\n",
      "\t* regulation of supramolecular fiber organ...\n",
      "\t* synapse organization\n",
      "\t* dendrite extension\n",
      "\t* regulation of transmembrane transporter ...\n",
      "\t* axon development\n",
      "\t* extracellular structure organization\n",
      "\t* regulation of system process\n",
      "\t* dendrite development\n",
      "\t* cellular response to organonitrogen comp...\n",
      "\t* regulation of Rho protein signal transdu...\n",
      "\t* cellular potassium ion transport\n",
      "\t* potassium ion transmembrane transport\n",
      "\t* muscle system process\n",
      "\t* negative regulation of sodium ion transm...\n",
      "\t* regulation of organelle organization\n",
      "\t* protein polymerization\n",
      "\t* regulation of cell morphogenesis\n",
      "\t* monovalent inorganic cation transport\n",
      "\t* regulation of cell projection assembly\n",
      "\t* regulation of molecular function\n",
      "\t* enzyme linked receptor protein signaling...\n",
      "\t* muscle contraction\n",
      "\t* phosphorus metabolic process\n",
      "\t* regulation of dendrite extension\n",
      "\t* regulation of ion transport\n",
      "\t* cellular response to nitrogen compound\n",
      "\t* regulation of response to stimulus\n",
      "\t* potassium ion transport\n",
      "\t* cell migration\n",
      "\t* regulation of cation transmembrane trans...\n",
      "\t* phosphate-containing compound metabolic ...\n",
      "\t* transmembrane receptor protein tyrosine ...\n",
      "\t* lipid translocation\n",
      "\t* behavior\n",
      "\t* microtubule cytoskeleton organization\n",
      "\t* cellular localization\n",
      "\t* regulation of locomotion\n",
      "\t* phospholipid translocation\n",
      "\t* negative regulation of sodium ion transp...\n",
      "\t* inositol lipid-mediated signaling\n",
      "\t* glutamate receptor signaling pathway\n",
      "\t* animal organ development\n",
      "\t* phosphatidylinositol-mediated signaling\n",
      "\t* regulation of postsynaptic membrane pote...\n",
      "\t* regulation of anatomical structure size\n",
      "\t* regulation of cilium assembly\n",
      "\t* regulation of cellular localization\n",
      "\t* calcium ion transmembrane transport\n",
      "\t* cell-substrate junction assembly\n",
      "\t* regulation of protein localization\n",
      "\t* regulation of cell motility\n",
      "\t* regulation of catalytic activity\n",
      "\t* negative regulation of protein depolymer...\n",
      "\t* axon guidance\n",
      "\t* cell junction assembly\n",
      "\t* regulation of cytoskeleton organization\n",
      "\t* regulation of protein depolymerization\n",
      "\t* neuron projection guidance\n",
      "\t* calcium ion transport\n",
      "\t* negative regulation of supramolecular fi...\n",
      "\t* protein localization\n",
      "\t* regulation of actin filament organizatio...\n",
      "\t* axoneme assembly\n",
      "\t* signaling\n",
      "\t* cellular response to caffeine\n",
      "\t* regulation of cation channel activity\n",
      "\t* cellular response to purine-containing c...\n",
      "\t* microtubule bundle formation\n",
      "\t* multicellular organismal signaling\n",
      "\t* neuron projection extension\n",
      "\t* regulation of cell-matrix adhesion\n",
      "\t* cell surface receptor signaling pathway ...\n",
      "\t* chemical synaptic transmission, postsyna...\n",
      "\t* developmental growth involved in morphog...\n",
      "\t* regulation of cell migration\n",
      "\t* excitatory postsynaptic potential\n",
      "\t* divalent metal ion transport\n",
      "\t* cell-substrate adherens junction assembl...\n",
      "\t* focal adhesion assembly\n",
      "\t* positive regulation of cell projection o...\n",
      "\t* growth\n",
      "\t* cell communication\n",
      "\t* regulation of actin filament-based proce...\n",
      "\t* negative regulation of cytoskeleton orga...\n",
      "\t* regulation of cellular component size\n",
      "\t* adherens junction assembly\n",
      "\t* retina layer formation\n",
      "\t* regulation of developmental process\n",
      "\t* divalent inorganic cation transport\n",
      "\t* negative regulation of protein complex d...\n",
      "\t* Rho protein signal transduction\n",
      "\t* protein depolymerization\n",
      "\t* regulation of actin cytoskeleton organiz...\n",
      "\t* brain development\n",
      "\t* cellular response to stimulus\n",
      "\t* head development\n",
      "\t* regulation of metal ion transport\n",
      "\t* regulation of cell differentiation\n",
      "\t* regulation of synaptic plasticity\n",
      "\t* vesicle cytoskeletal trafficking\n",
      "\t* regulation of hydrolase activity\n",
      "\t* circulatory system process\n",
      "\t* nuclear division\n",
      "\t* cardiac conduction\n",
      "\t* regulation of multicellular organismal d...\n",
      "\t* blood circulation\n",
      "\t* dendrite morphogenesis\n",
      "\t* developmental cell growth\n",
      "\t* epithelial cell morphogenesis\n",
      "\t* chromosome separation\n",
      "\t* regulation of platelet aggregation\n",
      "\t* phospholipid transport\n",
      "\t* positive regulation of cell-matrix adhes...\n",
      "\t* regulation of neurotransmitter secretion\n",
      "\t* bone development\n",
      "\t* regulation of cell-substrate adhesion\n",
      "\t* cilium or flagellum-dependent cell motil...\n",
      "\t* cilium-dependent cell motility\n",
      "\t* positive regulation of cellular componen...\n",
      "\t* regulation of blood circulation\n",
      "\t* response to drug\n",
      "\t* cellular response to drug\n",
      "\t* positive regulation of neuron differenti...\n",
      "\t* establishment of vesicle localization\n",
      "\t* phosphatidylinositol 3-kinase signaling\n",
      "\t* vesicle localization\n",
      "\t* organelle fission\n",
      "\t* regulation of protein complex disassembl...\n",
      "\t* membrane depolarization\n",
      "\t* positive regulation of dendrite extensio...\n",
      "\t* positive regulation of calcium ion trans...\n",
      "\t* heart process\n",
      "\t* sensory perception of mechanical stimulu...\n",
      "\t* actin filament-based movement\n",
      "\t* anion transport\n",
      "\t* locomotory behavior\n",
      "\t* protein phosphorylation\n",
      "\t* negative regulation of synaptic transmis...\n",
      "\t* regulation of calcium ion transmembrane ...\n",
      "\t* protein secretion\n",
      "\t* heart contraction\n",
      "\t* muscle structure development\n",
      "\t* epithelial cell development\n",
      "\t* adult behavior\n",
      "\t* kidney development\n",
      "\t* negative regulation of platelet aggregat...\n",
      "\t* heparan sulfate proteoglycan biosyntheti...\n",
      "\t* positive regulation of neurogenesis\n",
      "\t* contractile actin filament bundle assemb...\n",
      "\t* stress fiber assembly\n",
      "\t* response to wounding\n",
      "\t* cell-cell junction organization\n",
      "\t* regulation of protein complex assembly\n",
      "\t* regulation of protein polymerization\n",
      "\t* response to organonitrogen compound\n",
      "\t* regulation of organelle assembly\n",
      "\t* regulation of platelet activation\n",
      "\t* positive regulation of cell development\n",
      "\t* wound healing\n",
      "\t* establishment or maintenance of cell pol...\n",
      "\t* learning\n",
      "\t* regulation of developmental growth\n",
      "\t* cytoskeleton-dependent intracellular tra...\n",
      "\t* negative regulation of protein complex a...\n",
      "\t* positive regulation of kinase activity\n",
      "\t* retina homeostasis\n",
      "\t* actin filament depolymerization\n",
      "\t* positive regulation of neuron projection...\n",
      "\t* tissue development\n",
      "\t* collagen fibril organization\n",
      "\t* response to stimulus\n",
      "\t* phosphorylation\n",
      "\t* visual system development\n",
      "\t* regulation of potassium ion transmembran...\n",
      "\t* neurotransmitter secretion\n",
      "\t* sperm axoneme assembly\n",
      "\t* regulation of glutamate receptor signali...\n",
      "\t* signal release\n",
      "\t* meiotic nuclear division\n",
      "\t* muscle cell development\n",
      "\t* signal release from synapse\n",
      "\t* organelle localization\n",
      "\t* regulation of membrane lipid distributio...\n",
      "\t* glycerolipid biosynthetic process\n",
      "\t* establishment of organelle localization\n",
      "\t* positive regulation of nervous system de...\n",
      "\t* eye development\n",
      "\t* heparan sulfate proteoglycan metabolic p...\n",
      "\t* response to ATP\n",
      "\t* positive regulation of multicellular org...\n",
      "\t* neurotransmitter transport\n",
      "\t* negative regulation of cellular componen...\n",
      "\t* establishment of cell polarity\n",
      "\t* retina development in camera-type eye\n",
      "\t* neuron migration\n",
      "\t* response to purine-containing compound\n",
      "\t* regulation of protein secretion\n"
     ]
    }
   ],
   "source": [
    "for term in unique_terms_CB: print(f\"\\t* {term}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Term</th>\n",
       "      <th>Annotated</th>\n",
       "      <th>Significant</th>\n",
       "      <th>Expected</th>\n",
       "      <th>Fisher Test</th>\n",
       "      <th>Significant Genes</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO.ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GO:0120036</th>\n",
       "      <td>plasma membrane bounded cell projection ...</td>\n",
       "      <td>1559</td>\n",
       "      <td>343</td>\n",
       "      <td>196.02</td>\n",
       "      <td>1.200000e-27</td>\n",
       "      <td>1002:10083:10090:10100:1012:10121:10257:10300:...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0030030</th>\n",
       "      <td>cell projection organization</td>\n",
       "      <td>1598</td>\n",
       "      <td>349</td>\n",
       "      <td>200.92</td>\n",
       "      <td>1.700000e-27</td>\n",
       "      <td>1002:10083:10090:10100:1012:10121:10257:10300:...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0016043</th>\n",
       "      <td>cellular component organization</td>\n",
       "      <td>6650</td>\n",
       "      <td>1053</td>\n",
       "      <td>836.14</td>\n",
       "      <td>5.100000e-23</td>\n",
       "      <td>10000:1002:1004:10049:10056:10079:1008:10082:1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0048468</th>\n",
       "      <td>cell development</td>\n",
       "      <td>2133</td>\n",
       "      <td>409</td>\n",
       "      <td>268.19</td>\n",
       "      <td>1.500000e-20</td>\n",
       "      <td>10019:1002:10083:10090:10100:1021:10300:10319:...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0071840</th>\n",
       "      <td>cellular component organization or bioge...</td>\n",
       "      <td>6845</td>\n",
       "      <td>1065</td>\n",
       "      <td>860.66</td>\n",
       "      <td>1.800000e-20</td>\n",
       "      <td>10000:1002:1004:10049:10056:10079:1008:10082:1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0030010</th>\n",
       "      <td>establishment of cell polarity</td>\n",
       "      <td>139</td>\n",
       "      <td>31</td>\n",
       "      <td>17.48</td>\n",
       "      <td>9.600000e-04</td>\n",
       "      <td>10090:10806:10890:143098:1794:23094:23150:2318...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0060041</th>\n",
       "      <td>retina development in camera-type eye</td>\n",
       "      <td>139</td>\n",
       "      <td>31</td>\n",
       "      <td>17.48</td>\n",
       "      <td>9.600000e-04</td>\n",
       "      <td>10083:10319:1826:2055:221322:221935:22882:2336...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0001764</th>\n",
       "      <td>neuron migration</td>\n",
       "      <td>145</td>\n",
       "      <td>32</td>\n",
       "      <td>18.23</td>\n",
       "      <td>9.700000e-04</td>\n",
       "      <td>10718:10806:116983:153090:161582:1952:2066:221...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0014074</th>\n",
       "      <td>response to purine-containing compound</td>\n",
       "      <td>145</td>\n",
       "      <td>32</td>\n",
       "      <td>18.23</td>\n",
       "      <td>9.700000e-04</td>\n",
       "      <td>10518:1261:1373:159:1791:2230:2889:3488:351:37...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0050708</th>\n",
       "      <td>regulation of protein secretion</td>\n",
       "      <td>282</td>\n",
       "      <td>54</td>\n",
       "      <td>35.46</td>\n",
       "      <td>9.900000e-04</td>\n",
       "      <td>10497:10972:11069:111:134957:1803:1814:252983:...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>318 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                   Term  Annotated  \\\n",
       "GO.ID                                                                \n",
       "GO:0120036  plasma membrane bounded cell projection ...       1559   \n",
       "GO:0030030                 cell projection organization       1598   \n",
       "GO:0016043              cellular component organization       6650   \n",
       "GO:0048468                             cell development       2133   \n",
       "GO:0071840  cellular component organization or bioge...       6845   \n",
       "...                                                 ...        ...   \n",
       "GO:0030010               establishment of cell polarity        139   \n",
       "GO:0060041        retina development in camera-type eye        139   \n",
       "GO:0001764                             neuron migration        145   \n",
       "GO:0014074       response to purine-containing compound        145   \n",
       "GO:0050708              regulation of protein secretion        282   \n",
       "\n",
       "            Significant  Expected   Fisher Test  \\\n",
       "GO.ID                                             \n",
       "GO:0120036          343    196.02  1.200000e-27   \n",
       "GO:0030030          349    200.92  1.700000e-27   \n",
       "GO:0016043         1053    836.14  5.100000e-23   \n",
       "GO:0048468          409    268.19  1.500000e-20   \n",
       "GO:0071840         1065    860.66  1.800000e-20   \n",
       "...                 ...       ...           ...   \n",
       "GO:0030010           31     17.48  9.600000e-04   \n",
       "GO:0060041           31     17.48  9.600000e-04   \n",
       "GO:0001764           32     18.23  9.700000e-04   \n",
       "GO:0014074           32     18.23  9.700000e-04   \n",
       "GO:0050708           54     35.46  9.900000e-04   \n",
       "\n",
       "                                            Significant Genes  \n",
       "GO.ID                                                          \n",
       "GO:0120036  1002:10083:10090:10100:1012:10121:10257:10300:...  \n",
       "GO:0030030  1002:10083:10090:10100:1012:10121:10257:10300:...  \n",
       "GO:0016043  10000:1002:1004:10049:10056:10079:1008:10082:1...  \n",
       "GO:0048468  10019:1002:10083:10090:10100:1021:10300:10319:...  \n",
       "GO:0071840  10000:1002:1004:10049:10056:10079:1008:10082:1...  \n",
       "...                                                       ...  \n",
       "GO:0030010  10090:10806:10890:143098:1794:23094:23150:2318...  \n",
       "GO:0060041  10083:10319:1826:2055:221322:221935:22882:2336...  \n",
       "GO:0001764  10718:10806:116983:153090:161582:1952:2066:221...  \n",
       "GO:0014074  10518:1261:1373:159:1791:2230:2889:3488:351:37...  \n",
       "GO:0050708  10497:10972:11069:111:134957:1803:1814:252983:...  \n",
       "\n",
       "[318 rows x 6 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "biological_processes_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### GO: **Celular Location**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "celular_location_df.shape: (123, 6)\n",
      "celular_location_df.columns: ['Term', 'Annotated', 'Significant', 'Expected', 'Fisher Test', 'Significant Genes']\n"
     ]
    }
   ],
   "source": [
    "celular_location_df = pd.read_csv(f\"{input_folder}/CC_enriched_genes_entrez.txt.tsv\", sep=\"\\t\", index_col=0, header=\"infer\")\n",
    "print(f\"celular_location_df.shape: {celular_location_df.shape}\")\n",
    "print(f\"celular_location_df.columns: {list(celular_location_df.columns)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of unique_terms: 122\n"
     ]
    }
   ],
   "source": [
    "unique_terms_CC = pd.unique(celular_location_df[\"Term\"])\n",
    "print(f\"Number of unique_terms: {len(unique_terms_CC)}\")\n",
    "# for unique_term in unique_terms_CC: print(unique_term)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Term</th>\n",
       "      <th>Annotated</th>\n",
       "      <th>Significant</th>\n",
       "      <th>Expected</th>\n",
       "      <th>Fisher Test</th>\n",
       "      <th>Significant Genes</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO.ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GO:0005737</th>\n",
       "      <td>cytoplasm</td>\n",
       "      <td>11890</td>\n",
       "      <td>1785</td>\n",
       "      <td>1501.93</td>\n",
       "      <td>&lt; 1e-30</td>\n",
       "      <td>10000:10010:10019:10020:10024:10026:100271849:...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0042995</th>\n",
       "      <td>cell projection</td>\n",
       "      <td>2267</td>\n",
       "      <td>470</td>\n",
       "      <td>286.37</td>\n",
       "      <td>&lt; 1e-30</td>\n",
       "      <td>10083:1012:1021:10300:10400:10458:10497:10518:...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0044444</th>\n",
       "      <td>cytoplasmic part</td>\n",
       "      <td>9986</td>\n",
       "      <td>1518</td>\n",
       "      <td>1261.42</td>\n",
       "      <td>7.2e-29</td>\n",
       "      <td>10010:10019:10020:10026:100271849:100271927:10...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0120025</th>\n",
       "      <td>plasma membrane bounded cell projection</td>\n",
       "      <td>2167</td>\n",
       "      <td>445</td>\n",
       "      <td>273.73</td>\n",
       "      <td>1.9e-28</td>\n",
       "      <td>10083:1012:1021:10300:10400:10458:10497:10518:...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0044463</th>\n",
       "      <td>cell projection part</td>\n",
       "      <td>1509</td>\n",
       "      <td>327</td>\n",
       "      <td>190.62</td>\n",
       "      <td>1.1e-24</td>\n",
       "      <td>10083:10300:10400:10497:10518:1063:10642:10755...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0033268</th>\n",
       "      <td>node of Ranvier</td>\n",
       "      <td>14</td>\n",
       "      <td>7</td>\n",
       "      <td>1.77</td>\n",
       "      <td>0.00078</td>\n",
       "      <td>1739:23114:288:3786:4653:57731:6900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0099513</th>\n",
       "      <td>polymeric cytoskeletal fiber</td>\n",
       "      <td>747</td>\n",
       "      <td>124</td>\n",
       "      <td>94.36</td>\n",
       "      <td>0.00078</td>\n",
       "      <td>10300:10611:10783:11043:11076:11127:114791:127...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0005783</th>\n",
       "      <td>endoplasmic reticulum</td>\n",
       "      <td>1960</td>\n",
       "      <td>293</td>\n",
       "      <td>247.59</td>\n",
       "      <td>0.00079</td>\n",
       "      <td>10026:10148:10299:10313:10345:10396:10400:1040...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0042734</th>\n",
       "      <td>presynaptic membrane</td>\n",
       "      <td>137</td>\n",
       "      <td>31</td>\n",
       "      <td>17.31</td>\n",
       "      <td>0.00081</td>\n",
       "      <td>1008:10497:1272:134957:1949:2066:22854:22859:2...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0032420</th>\n",
       "      <td>stereocilium</td>\n",
       "      <td>55</td>\n",
       "      <td>16</td>\n",
       "      <td>6.95</td>\n",
       "      <td>0.00093</td>\n",
       "      <td>10083:10518:125336:2059:25861:583:64072:7399:7...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>123 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               Term  Annotated  Significant  \\\n",
       "GO.ID                                                                         \n",
       "GO:0005737                                cytoplasm      11890         1785   \n",
       "GO:0042995                          cell projection       2267          470   \n",
       "GO:0044444                         cytoplasmic part       9986         1518   \n",
       "GO:0120025  plasma membrane bounded cell projection       2167          445   \n",
       "GO:0044463                     cell projection part       1509          327   \n",
       "...                                             ...        ...          ...   \n",
       "GO:0033268                          node of Ranvier         14            7   \n",
       "GO:0099513             polymeric cytoskeletal fiber        747          124   \n",
       "GO:0005783                    endoplasmic reticulum       1960          293   \n",
       "GO:0042734                     presynaptic membrane        137           31   \n",
       "GO:0032420                             stereocilium         55           16   \n",
       "\n",
       "            Expected Fisher Test  \\\n",
       "GO.ID                              \n",
       "GO:0005737   1501.93     < 1e-30   \n",
       "GO:0042995    286.37     < 1e-30   \n",
       "GO:0044444   1261.42     7.2e-29   \n",
       "GO:0120025    273.73     1.9e-28   \n",
       "GO:0044463    190.62     1.1e-24   \n",
       "...              ...         ...   \n",
       "GO:0033268      1.77     0.00078   \n",
       "GO:0099513     94.36     0.00078   \n",
       "GO:0005783    247.59     0.00079   \n",
       "GO:0042734     17.31     0.00081   \n",
       "GO:0032420      6.95     0.00093   \n",
       "\n",
       "                                            Significant Genes  \n",
       "GO.ID                                                          \n",
       "GO:0005737  10000:10010:10019:10020:10024:10026:100271849:...  \n",
       "GO:0042995  10083:1012:1021:10300:10400:10458:10497:10518:...  \n",
       "GO:0044444  10010:10019:10020:10026:100271849:100271927:10...  \n",
       "GO:0120025  10083:1012:1021:10300:10400:10458:10497:10518:...  \n",
       "GO:0044463  10083:10300:10400:10497:10518:1063:10642:10755...  \n",
       "...                                                       ...  \n",
       "GO:0033268                1739:23114:288:3786:4653:57731:6900  \n",
       "GO:0099513  10300:10611:10783:11043:11076:11127:114791:127...  \n",
       "GO:0005783  10026:10148:10299:10313:10345:10396:10400:1040...  \n",
       "GO:0042734  1008:10497:1272:134957:1949:2066:22854:22859:2...  \n",
       "GO:0032420  10083:10518:125336:2059:25861:583:64072:7399:7...  \n",
       "\n",
       "[123 rows x 6 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "celular_location_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "membrane_counter = 0\n",
    "membrane_related = [\"membrane\", \"cytoskeleton\", \"periphery\"]\n",
    "for term in unique_terms_CC: \n",
    "    # print(f\"\\t* {term}\")\n",
    "    for word in membrane_related: \n",
    "        if word in term: membrane_counter +=1\n",
    "# print(f\"membrane_counter: {membrane_counter}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### GO: **Molecular Function**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "molecular_function_df.shape: (92, 6)\n",
      "molecular_function_df.columns: ['Term', 'Annotated', 'Significant', 'Expected', 'Fisher Test', 'Significant Genes']\n"
     ]
    }
   ],
   "source": [
    "molecular_function_df = pd.read_csv(f\"{input_folder}/MF_enriched_genes_entrez.txt.tsv\", sep=\"\\t\", index_col=0, header=\"infer\")\n",
    "print(f\"molecular_function_df.shape: {molecular_function_df.shape}\")\n",
    "print(f\"molecular_function_df.columns: {list(molecular_function_df.columns)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of unique_terms: 89\n"
     ]
    }
   ],
   "source": [
    "unique_terms_MF = pd.unique(molecular_function_df[\"Term\"])\n",
    "print(f\"Number of unique_terms: {len(unique_terms_MF)}\")\n",
    "# for unique_term in unique_terms_MF: print(unique_term)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# for term in unique_terms_MF: print(f\"\\t* {term}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Term</th>\n",
       "      <th>Annotated</th>\n",
       "      <th>Significant</th>\n",
       "      <th>Expected</th>\n",
       "      <th>Fisher Test</th>\n",
       "      <th>Significant Genes</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO.ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GO:0008092</th>\n",
       "      <td>cytoskeletal protein binding</td>\n",
       "      <td>985</td>\n",
       "      <td>248</td>\n",
       "      <td>130.35</td>\n",
       "      <td>2.100000e-25</td>\n",
       "      <td>10083:10300:10391:10398:10458:10580:10611:1063...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0043168</th>\n",
       "      <td>anion binding</td>\n",
       "      <td>2844</td>\n",
       "      <td>535</td>\n",
       "      <td>376.35</td>\n",
       "      <td>3.100000e-20</td>\n",
       "      <td>10000:10020:10026:100271927:100506084:10056:10...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0003779</th>\n",
       "      <td>actin binding</td>\n",
       "      <td>441</td>\n",
       "      <td>123</td>\n",
       "      <td>58.36</td>\n",
       "      <td>1.300000e-16</td>\n",
       "      <td>10083:10391:10580:10611:10755:11170:114793:119...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0043167</th>\n",
       "      <td>ion binding</td>\n",
       "      <td>6293</td>\n",
       "      <td>1010</td>\n",
       "      <td>832.76</td>\n",
       "      <td>5.500000e-16</td>\n",
       "      <td>10000:10010:100131539:1002:10020:10026:1002719...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0005543</th>\n",
       "      <td>phospholipid binding</td>\n",
       "      <td>456</td>\n",
       "      <td>115</td>\n",
       "      <td>60.34</td>\n",
       "      <td>2.400000e-12</td>\n",
       "      <td>10026:100271927:10400:10497:1071:11188:11199:1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0042578</th>\n",
       "      <td>phosphoric ester hydrolase activity</td>\n",
       "      <td>372</td>\n",
       "      <td>71</td>\n",
       "      <td>49.23</td>\n",
       "      <td>8.500000e-04</td>\n",
       "      <td>10076:11122:1131:151742:1739:1846:1909:2203:22...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0004674</th>\n",
       "      <td>protein serine/threonine kinase activity</td>\n",
       "      <td>391</td>\n",
       "      <td>74</td>\n",
       "      <td>51.74</td>\n",
       "      <td>8.500000e-04</td>\n",
       "      <td>10000:1018:1021:10420:10746:10783:11214:139728...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0015077</th>\n",
       "      <td>monovalent inorganic cation transmembran...</td>\n",
       "      <td>367</td>\n",
       "      <td>70</td>\n",
       "      <td>48.57</td>\n",
       "      <td>9.300000e-04</td>\n",
       "      <td>10479:11280:1183:131096:150159:153201:157855:1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0015276</th>\n",
       "      <td>ligand-gated ion channel activity</td>\n",
       "      <td>138</td>\n",
       "      <td>32</td>\n",
       "      <td>18.26</td>\n",
       "      <td>9.500000e-04</td>\n",
       "      <td>1143:116443:1258:1261:200959:2560:2567:2741:27...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0022834</th>\n",
       "      <td>ligand-gated channel activity</td>\n",
       "      <td>138</td>\n",
       "      <td>32</td>\n",
       "      <td>18.26</td>\n",
       "      <td>9.500000e-04</td>\n",
       "      <td>1143:116443:1258:1261:200959:2560:2567:2741:27...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>92 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                   Term  Annotated  \\\n",
       "GO.ID                                                                \n",
       "GO:0008092                 cytoskeletal protein binding        985   \n",
       "GO:0043168                                anion binding       2844   \n",
       "GO:0003779                                actin binding        441   \n",
       "GO:0043167                                  ion binding       6293   \n",
       "GO:0005543                         phospholipid binding        456   \n",
       "...                                                 ...        ...   \n",
       "GO:0042578          phosphoric ester hydrolase activity        372   \n",
       "GO:0004674     protein serine/threonine kinase activity        391   \n",
       "GO:0015077  monovalent inorganic cation transmembran...        367   \n",
       "GO:0015276            ligand-gated ion channel activity        138   \n",
       "GO:0022834                ligand-gated channel activity        138   \n",
       "\n",
       "            Significant  Expected   Fisher Test  \\\n",
       "GO.ID                                             \n",
       "GO:0008092          248    130.35  2.100000e-25   \n",
       "GO:0043168          535    376.35  3.100000e-20   \n",
       "GO:0003779          123     58.36  1.300000e-16   \n",
       "GO:0043167         1010    832.76  5.500000e-16   \n",
       "GO:0005543          115     60.34  2.400000e-12   \n",
       "...                 ...       ...           ...   \n",
       "GO:0042578           71     49.23  8.500000e-04   \n",
       "GO:0004674           74     51.74  8.500000e-04   \n",
       "GO:0015077           70     48.57  9.300000e-04   \n",
       "GO:0015276           32     18.26  9.500000e-04   \n",
       "GO:0022834           32     18.26  9.500000e-04   \n",
       "\n",
       "                                            Significant Genes  \n",
       "GO.ID                                                          \n",
       "GO:0008092  10083:10300:10391:10398:10458:10580:10611:1063...  \n",
       "GO:0043168  10000:10020:10026:100271927:100506084:10056:10...  \n",
       "GO:0003779  10083:10391:10580:10611:10755:11170:114793:119...  \n",
       "GO:0043167  10000:10010:100131539:1002:10020:10026:1002719...  \n",
       "GO:0005543  10026:100271927:10400:10497:1071:11188:11199:1...  \n",
       "...                                                       ...  \n",
       "GO:0042578  10076:11122:1131:151742:1739:1846:1909:2203:22...  \n",
       "GO:0004674  10000:1018:1021:10420:10746:10783:11214:139728...  \n",
       "GO:0015077  10479:11280:1183:131096:150159:153201:157855:1...  \n",
       "GO:0015276  1143:116443:1258:1261:200959:2560:2567:2741:27...  \n",
       "GO:0022834  1143:116443:1258:1261:200959:2560:2567:2741:27...  \n",
       "\n",
       "[92 rows x 6 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "molecular_function_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **Quick trial of Text Mining clustering**\n",
    "\n",
    "Reference: https://medium.com/mlearning-ai/text-clustering-with-tf-idf-in-python-c94cd26a31e7"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "def preprocess_text(text: str, remove_stopwords: bool) -> str:\n",
    "    \n",
    "    \"\"\"\n",
    "    This function cleans the input text by\n",
    "    - removing links\n",
    "    - removing special chars\n",
    "    - removing numbers\n",
    "    - removing stopwords\n",
    "    - transforming in lower case\n",
    "    - removing excessive whitespaces\n",
    "    Arguments:\n",
    "        text (str): text to clean\n",
    "        remove_stopwords (bool): remove stopwords or not\n",
    "    Returns:\n",
    "        str: cleaned text\"\"\"\n",
    "        \n",
    "    # remove links\n",
    "    text = re.sub(r\"http\\S+\", \"\", text)\n",
    "    # remove numbers and special chars\n",
    "    text = re.sub(\"[^A-Za-z]+\", \" \", text)\n",
    "    # remove stopwords\n",
    "    if remove_stopwords:\n",
    "        # 1. creates tokens\n",
    "        tokens = nltk.word_tokenize(text)\n",
    "        # 2. checks if token is a stopword and removes it\n",
    "        tokens = [w for w in tokens if not w.lower() in stopwords.words(\"english\")]\n",
    "        # 3. joins all tokens again\n",
    "        text = \" \".join(tokens)\n",
    "    # returns cleaned text\n",
    "    text = text.lower().strip()\n",
    "    return text\n",
    "  \n",
    "def get_top_keywords(n_terms):\n",
    "    \"\"\"This function returns the keywords for each centroid of the KMeans\"\"\"\n",
    "    df = pd.DataFrame(X.todense()).groupby(clusters).mean() # groups tf idf vector per cluster\n",
    "    terms = vectorizer.get_feature_names_out() # access to tf idf terms\n",
    "    for i,r in df.iterrows():\n",
    "        print('\\nCluster {}'.format(i))\n",
    "        print(','.join([terms[t] for t in np.argsort(r)[-n_terms:]])) # for each row of the dataframe, find the n terms that have the highest tf idf score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "ename": "ModuleNotFoundError",
     "evalue": "No module named 'nltk'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)",
      "Input \u001b[0;32mIn [16]\u001b[0m, in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     12\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mre\u001b[39;00m\n\u001b[1;32m     13\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mstring\u001b[39;00m\n\u001b[0;32m---> 14\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mnltk\u001b[39;00m\n\u001b[1;32m     15\u001b[0m nltk\u001b[38;5;241m.\u001b[39mdownload(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mpunkt\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[1;32m     16\u001b[0m nltk\u001b[38;5;241m.\u001b[39mdownload(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mstopwords\u001b[39m\u001b[38;5;124m'\u001b[39m)\n",
      "\u001b[0;31mModuleNotFoundError\u001b[0m: No module named 'nltk'"
     ]
    }
   ],
   "source": [
    "# import required sklearn libs\n",
    "from sklearn.datasets import fetch_20newsgroups\n",
    "from sklearn.feature_extraction.text import TfidfVectorizer\n",
    "from sklearn.cluster import KMeans\n",
    "from sklearn.decomposition import PCA\n",
    "\n",
    "# import other required libs\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "# string manipulation libs\n",
    "import re\n",
    "import string\n",
    "import nltk\n",
    "nltk.download('punkt')\n",
    "nltk.download('stopwords')\n",
    "from nltk.corpus import stopwords\n",
    "\n",
    "# viz libs\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "df = pd.DataFrame(biological_processes_df[\"Term\"], columns=[\"Term\"])\n",
    "df['cleaned'] = df['Term'].apply(lambda x: preprocess_text(x, remove_stopwords=True))\n",
    "\n",
    "# initialize vectorizer\n",
    "vectorizer = TfidfVectorizer(sublinear_tf=True, min_df=5, max_df=0.95)\n",
    "# fit_transform applies TF-IDF to clean texts - we save the array of vectors in X\n",
    "X = vectorizer.fit_transform(df['cleaned'])\n",
    "\n",
    "# initialize KMeans with 3 clusters\n",
    "kmeans = KMeans(n_clusters=5, random_state=42)\n",
    "kmeans.fit(X)\n",
    "clusters = kmeans.labels_\n",
    "biological_processes_df[\"cluster\"] = clusters\n",
    "\n",
    "# initialize PCA with 2 components\n",
    "pca = PCA(n_components=2, random_state=42)\n",
    "# pass X to the pca\n",
    "pca_vecs = pca.fit_transform(X.toarray())\n",
    "# save the two dimensions in x0 and x1\n",
    "x0 = pca_vecs[:, 0]\n",
    "x1 = pca_vecs[:, 1]\n",
    "\n",
    "# assign clusters and PCA vectors to columns in the original dataframe\n",
    "df['cluster'] = clusters\n",
    "df['x0'] = x0\n",
    "df['x1'] = x1\n",
    "\n",
    "# cluster_map = {0: \"membrane organization\", 1: \"neurogenesis\", 2: \"cell organization\"} # mapping found through get_top_keywords\n",
    "# df['cluster'] = df['cluster'].map(cluster_map)\n",
    "\n",
    "# set image size\n",
    "plt.figure(figsize=(12, 7))\n",
    "# set title\n",
    "plt.title(\"Grouping TF-IDF + KMeans 20newsgroup\", fontdict={\"fontsize\": 18})\n",
    "# set axes names\n",
    "plt.xlabel(\"X0\", fontdict={\"fontsize\": 16})\n",
    "plt.ylabel(\"X1\", fontdict={\"fontsize\": 16})\n",
    "#  create scatter plot with seaborn, where hue is the class used to group the data\n",
    "sns.scatterplot(data=df, x='x0', y='x1', hue='cluster', palette=\"viridis\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# for cluster in clusters:\n",
    "#     print(f\"cluster: {cluster}\")\n",
    "#     df = biological_processes_df[biological_processes_df[\"cluster\"] == cluster]\n",
    "#     print(df[\"Term\"])\n",
    "#     print()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### (2) **UCSC annotation** (original)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ucsc_annotation_df.shape: (8066, 4)\n"
     ]
    }
   ],
   "source": [
    "input_folder=\"my_results\"  # Update input folder\n",
    "ucsc_annotation_df = pd.read_csv(f\"{input_folder}/UCSC_common_segments_annotated.bed\", sep=\"\\t\", index_col=None, header=None)\n",
    "print(f\"ucsc_annotation_df.shape: {ucsc_annotation_df.shape}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NaN values in annotation: 4800 in 8066\n",
      "Number of different gene IDs: 2617\n"
     ]
    }
   ],
   "source": [
    "# NaN values in annotation column\n",
    "print(f\"NaN values in annotation: {ucsc_annotation_df[3].isnull().sum()} in {ucsc_annotation_df.shape[0]}\")\n",
    "\n",
    "# Change NaN values for desired value\n",
    "desired_value=\"NoGene\"\n",
    "ucsc_annotation_df = ucsc_annotation_df.fillna(desired_value, axis=1)\n",
    "\n",
    "# Unique identifiers\n",
    "unique_ids = pd.unique(ucsc_annotation_df[3])\n",
    "unique_ids = np.delete(unique_ids, np.where(unique_ids == \"NoGene\"))\n",
    "print(f\"Number of different gene IDs: {len(unique_ids)}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### (3) **UCSC annotation** (coding)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Pendiente comprender significado de dos archivos."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ucsc_annotation_codingComplete.shape: (21461, 12)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>66999835</td>\n",
       "      <td>67216822</td>\n",
       "      <td>NM_032291.4</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>67000041</td>\n",
       "      <td>67208778</td>\n",
       "      <td>0</td>\n",
       "      <td>25</td>\n",
       "      <td>216,64,25,72,57,55,176,12,12,25,52,86,93,75,50...</td>\n",
       "      <td>0,91694,98917,101791,105624,108657,109391,1263...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>8378173</td>\n",
       "      <td>8404225</td>\n",
       "      <td>NM_001080397.3</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>8384389</td>\n",
       "      <td>8404073</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>73,421,93,225,728,154,177,206,419,</td>\n",
       "      <td>0,6192,7184,7704,12095,17323,19702,21379,25633,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>16767218</td>\n",
       "      <td>16786573</td>\n",
       "      <td>NM_018090.5</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>16767256</td>\n",
       "      <td>16785385</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>130,101,105,82,109,178,76,1237,</td>\n",
       "      <td>0,2908,7146,7336,8369,11114,15094,18118,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>25071937</td>\n",
       "      <td>25170812</td>\n",
       "      <td>NM_013943.3</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>25072044</td>\n",
       "      <td>25167428</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>179,110,126,107,182,3549,</td>\n",
       "      <td>0,52295,68647,81563,94413,95326,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>33546753</td>\n",
       "      <td>33589093</td>\n",
       "      <td>NM_052998.4</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>33547850</td>\n",
       "      <td>33585783</td>\n",
       "      <td>0</td>\n",
       "      <td>12</td>\n",
       "      <td>142,121,212,177,174,173,135,166,163,113,215,3449,</td>\n",
       "      <td>0,235,448,1025,2801,10897,12129,13395,15554,16...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21456</th>\n",
       "      <td>chr22_kb663609_alt</td>\n",
       "      <td>1</td>\n",
       "      <td>1783</td>\n",
       "      <td>NM_001002034.3</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>418</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1782,</td>\n",
       "      <td>0,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21457</th>\n",
       "      <td>chr22_kb663609_alt</td>\n",
       "      <td>2044</td>\n",
       "      <td>6629</td>\n",
       "      <td>NM_033318.5</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>2113</td>\n",
       "      <td>4407</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>255,141,1166,</td>\n",
       "      <td>0,2225,3419,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21458</th>\n",
       "      <td>chr22_kb663609_alt</td>\n",
       "      <td>7870</td>\n",
       "      <td>13117</td>\n",
       "      <td>NM_002490.6</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>8527</td>\n",
       "      <td>13089</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>789,116,167,</td>\n",
       "      <td>0,1534,5080,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21459</th>\n",
       "      <td>chr22_kb663609_alt</td>\n",
       "      <td>48825</td>\n",
       "      <td>53137</td>\n",
       "      <td>NM_000106.6</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>48900</td>\n",
       "      <td>53118</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>254,142,188,142,177,161,153,172,199,</td>\n",
       "      <td>0,352,948,1343,1675,2285,2534,3239,4113,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21460</th>\n",
       "      <td>chr7_gl000195_random</td>\n",
       "      <td>42937</td>\n",
       "      <td>86735</td>\n",
       "      <td>NM_001347681.2</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>44723</td>\n",
       "      <td>49117</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>1986,165,218,</td>\n",
       "      <td>0,6017,43580,</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>21461 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                         0         1         2               3   4  5   \\\n",
       "0                      chr1  66999835  67216822     NM_032291.4   0  +   \n",
       "1                      chr1   8378173   8404225  NM_001080397.3   0  +   \n",
       "2                      chr1  16767218  16786573     NM_018090.5   0  +   \n",
       "3                      chr1  25071937  25170812     NM_013943.3   0  +   \n",
       "4                      chr1  33546753  33589093     NM_052998.4   0  +   \n",
       "...                     ...       ...       ...             ...  .. ..   \n",
       "21456    chr22_kb663609_alt         1      1783  NM_001002034.3   0  +   \n",
       "21457    chr22_kb663609_alt      2044      6629     NM_033318.5   0  +   \n",
       "21458    chr22_kb663609_alt      7870     13117     NM_002490.6   0  -   \n",
       "21459    chr22_kb663609_alt     48825     53137     NM_000106.6   0  -   \n",
       "21460  chr7_gl000195_random     42937     86735  NM_001347681.2   0  -   \n",
       "\n",
       "             6         7   8   9   \\\n",
       "0      67000041  67208778   0  25   \n",
       "1       8384389   8404073   0   9   \n",
       "2      16767256  16785385   0   8   \n",
       "3      25072044  25167428   0   6   \n",
       "4      33547850  33585783   0  12   \n",
       "...         ...       ...  ..  ..   \n",
       "21456         1       418   0   1   \n",
       "21457      2113      4407   0   3   \n",
       "21458      8527     13089   0   3   \n",
       "21459     48900     53118   0   9   \n",
       "21460     44723     49117   0   3   \n",
       "\n",
       "                                                      10  \\\n",
       "0      216,64,25,72,57,55,176,12,12,25,52,86,93,75,50...   \n",
       "1                     73,421,93,225,728,154,177,206,419,   \n",
       "2                        130,101,105,82,109,178,76,1237,   \n",
       "3                              179,110,126,107,182,3549,   \n",
       "4      142,121,212,177,174,173,135,166,163,113,215,3449,   \n",
       "...                                                  ...   \n",
       "21456                                              1782,   \n",
       "21457                                      255,141,1166,   \n",
       "21458                                       789,116,167,   \n",
       "21459               254,142,188,142,177,161,153,172,199,   \n",
       "21460                                      1986,165,218,   \n",
       "\n",
       "                                                      11  \n",
       "0      0,91694,98917,101791,105624,108657,109391,1263...  \n",
       "1        0,6192,7184,7704,12095,17323,19702,21379,25633,  \n",
       "2               0,2908,7146,7336,8369,11114,15094,18118,  \n",
       "3                       0,52295,68647,81563,94413,95326,  \n",
       "4      0,235,448,1025,2801,10897,12129,13395,15554,16...  \n",
       "...                                                  ...  \n",
       "21456                                                 0,  \n",
       "21457                                       0,2225,3419,  \n",
       "21458                                       0,1534,5080,  \n",
       "21459           0,352,948,1343,1675,2285,2534,3239,4113,  \n",
       "21460                                      0,6017,43580,  \n",
       "\n",
       "[21461 rows x 12 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "input_folder=\"input/forExercise2\"  # Update input folder\n",
    "ucsc_annotation_codingComplete = pd.read_csv(f\"{input_folder}/protCodGenesComplete\", sep=\"\\t\", index_col=None, header=None)\n",
    "print(f\"ucsc_annotation_codingComplete.shape: {ucsc_annotation_codingComplete.shape}\")\n",
    "ucsc_annotation_codingComplete"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of different gene IDs: 19199\n"
     ]
    }
   ],
   "source": [
    "# Unique identifiers\n",
    "unique_ids = pd.unique(ucsc_annotation_codingComplete[3])\n",
    "print(f\"Number of different gene IDs: {len(unique_ids)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ucsc_annotation_codingExons.shape: (210136, 6)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>67000041</td>\n",
       "      <td>67000051</td>\n",
       "      <td>NM_032291.4_cds_0_0_chr1_67000042_f</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>67091529</td>\n",
       "      <td>67091593</td>\n",
       "      <td>NM_032291.4_cds_1_0_chr1_67091530_f</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>67098752</td>\n",
       "      <td>67098777</td>\n",
       "      <td>NM_032291.4_cds_2_0_chr1_67098753_f</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>67101626</td>\n",
       "      <td>67101698</td>\n",
       "      <td>NM_032291.4_cds_3_0_chr1_67101627_f</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>67105459</td>\n",
       "      <td>67105516</td>\n",
       "      <td>NM_032291.4_cds_4_0_chr1_67105460_f</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>210131</th>\n",
       "      <td>chr22_kb663609_alt</td>\n",
       "      <td>51359</td>\n",
       "      <td>51512</td>\n",
       "      <td>NM_000106.6_cds_6_0_chr22_kb663609_alt_51360_r</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>210132</th>\n",
       "      <td>chr22_kb663609_alt</td>\n",
       "      <td>52064</td>\n",
       "      <td>52236</td>\n",
       "      <td>NM_000106.6_cds_7_0_chr22_kb663609_alt_52065_r</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>210133</th>\n",
       "      <td>chr22_kb663609_alt</td>\n",
       "      <td>52938</td>\n",
       "      <td>53118</td>\n",
       "      <td>NM_000106.6_cds_8_0_chr22_kb663609_alt_52939_r</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>210134</th>\n",
       "      <td>chr7_gl000195_random</td>\n",
       "      <td>44723</td>\n",
       "      <td>44923</td>\n",
       "      <td>NM_001347681.2_cds_0_0_chr7_gl000195_random_44...</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>210135</th>\n",
       "      <td>chr7_gl000195_random</td>\n",
       "      <td>48954</td>\n",
       "      <td>49117</td>\n",
       "      <td>NM_001347681.2_cds_1_0_chr7_gl000195_random_48...</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>210136 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           0         1         2  \\\n",
       "0                       chr1  67000041  67000051   \n",
       "1                       chr1  67091529  67091593   \n",
       "2                       chr1  67098752  67098777   \n",
       "3                       chr1  67101626  67101698   \n",
       "4                       chr1  67105459  67105516   \n",
       "...                      ...       ...       ...   \n",
       "210131    chr22_kb663609_alt     51359     51512   \n",
       "210132    chr22_kb663609_alt     52064     52236   \n",
       "210133    chr22_kb663609_alt     52938     53118   \n",
       "210134  chr7_gl000195_random     44723     44923   \n",
       "210135  chr7_gl000195_random     48954     49117   \n",
       "\n",
       "                                                        3  4  5  \n",
       "0                     NM_032291.4_cds_0_0_chr1_67000042_f  0  +  \n",
       "1                     NM_032291.4_cds_1_0_chr1_67091530_f  0  +  \n",
       "2                     NM_032291.4_cds_2_0_chr1_67098753_f  0  +  \n",
       "3                     NM_032291.4_cds_3_0_chr1_67101627_f  0  +  \n",
       "4                     NM_032291.4_cds_4_0_chr1_67105460_f  0  +  \n",
       "...                                                   ... .. ..  \n",
       "210131     NM_000106.6_cds_6_0_chr22_kb663609_alt_51360_r  0  -  \n",
       "210132     NM_000106.6_cds_7_0_chr22_kb663609_alt_52065_r  0  -  \n",
       "210133     NM_000106.6_cds_8_0_chr22_kb663609_alt_52939_r  0  -  \n",
       "210134  NM_001347681.2_cds_0_0_chr7_gl000195_random_44...  0  -  \n",
       "210135  NM_001347681.2_cds_1_0_chr7_gl000195_random_48...  0  -  \n",
       "\n",
       "[210136 rows x 6 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "input_folder=\"input/forExercise2\"  # Update input folder\n",
    "ucsc_annotation_codingExons = pd.read_csv(f\"{input_folder}/protCodGenesONLYEXONS\", sep=\"\\t\", index_col=None, header=None)\n",
    "print(f\"ucsc_annotation_codingExons.shape: {ucsc_annotation_codingExons.shape}\")\n",
    "ucsc_annotation_codingExons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of different gene IDs: 210136\n"
     ]
    }
   ],
   "source": [
    "# Unique identifiers\n",
    "unique_ids = pd.unique(ucsc_annotation_codingExons[3])\n",
    "print(f\"Number of different gene IDs: {len(unique_ids)}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### (4) **UCSC annotation** (new in input, other BED features)\n",
    "\n",
    "See percentage criterion, based on -f option in `bedtools`, in summaryBiologicalInterpretation.ipynb.\n",
    "\n",
    "If we want a minimum overlap correspondent to the number of bases of the main CTCF DNA binding motif (17 bp) [**refs. 3 & 4**], we would need to know the average length of segments in the common_segments.bed file (...)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### **3UTR**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "utr3.shape: (22127, 6)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>67208778</td>\n",
       "      <td>67216822</td>\n",
       "      <td>NM_032291.4_utr3_24_0_chr1_67208779_f</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>8404073</td>\n",
       "      <td>8404225</td>\n",
       "      <td>NM_001080397.3_utr3_8_0_chr1_8404074_f</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>16785385</td>\n",
       "      <td>16786573</td>\n",
       "      <td>NM_018090.5_utr3_7_0_chr1_16785386_f</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>25167428</td>\n",
       "      <td>25170812</td>\n",
       "      <td>NM_013943.3_utr3_5_0_chr1_25167429_f</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>33585783</td>\n",
       "      <td>33589093</td>\n",
       "      <td>NM_052998.4_utr3_11_0_chr1_33585784_f</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22122</th>\n",
       "      <td>chr22_kb663609_alt</td>\n",
       "      <td>4407</td>\n",
       "      <td>4410</td>\n",
       "      <td>NM_033318.5_utr3_1_0_chr22_kb663609_alt_4408_f</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22123</th>\n",
       "      <td>chr22_kb663609_alt</td>\n",
       "      <td>5463</td>\n",
       "      <td>6629</td>\n",
       "      <td>NM_033318.5_utr3_2_0_chr22_kb663609_alt_5464_f</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22124</th>\n",
       "      <td>chr22_kb663609_alt</td>\n",
       "      <td>7870</td>\n",
       "      <td>8527</td>\n",
       "      <td>NM_002490.6_utr3_0_0_chr22_kb663609_alt_7871_r</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22125</th>\n",
       "      <td>chr22_kb663609_alt</td>\n",
       "      <td>48825</td>\n",
       "      <td>48900</td>\n",
       "      <td>NM_000106.6_utr3_0_0_chr22_kb663609_alt_48826_r</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22126</th>\n",
       "      <td>chr7_gl000195_random</td>\n",
       "      <td>42937</td>\n",
       "      <td>44723</td>\n",
       "      <td>NM_001347681.2_utr3_0_0_chr7_gl000195_random_4...</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>22127 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          0         1         2  \\\n",
       "0                      chr1  67208778  67216822   \n",
       "1                      chr1   8404073   8404225   \n",
       "2                      chr1  16785385  16786573   \n",
       "3                      chr1  25167428  25170812   \n",
       "4                      chr1  33585783  33589093   \n",
       "...                     ...       ...       ...   \n",
       "22122    chr22_kb663609_alt      4407      4410   \n",
       "22123    chr22_kb663609_alt      5463      6629   \n",
       "22124    chr22_kb663609_alt      7870      8527   \n",
       "22125    chr22_kb663609_alt     48825     48900   \n",
       "22126  chr7_gl000195_random     42937     44723   \n",
       "\n",
       "                                                       3  4  5  \n",
       "0                  NM_032291.4_utr3_24_0_chr1_67208779_f  0  +  \n",
       "1                 NM_001080397.3_utr3_8_0_chr1_8404074_f  0  +  \n",
       "2                   NM_018090.5_utr3_7_0_chr1_16785386_f  0  +  \n",
       "3                   NM_013943.3_utr3_5_0_chr1_25167429_f  0  +  \n",
       "4                  NM_052998.4_utr3_11_0_chr1_33585784_f  0  +  \n",
       "...                                                  ... .. ..  \n",
       "22122     NM_033318.5_utr3_1_0_chr22_kb663609_alt_4408_f  0  +  \n",
       "22123     NM_033318.5_utr3_2_0_chr22_kb663609_alt_5464_f  0  +  \n",
       "22124     NM_002490.6_utr3_0_0_chr22_kb663609_alt_7871_r  0  -  \n",
       "22125    NM_000106.6_utr3_0_0_chr22_kb663609_alt_48826_r  0  -  \n",
       "22126  NM_001347681.2_utr3_0_0_chr7_gl000195_random_4...  0  -  \n",
       "\n",
       "[22127 rows x 6 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "input_folder=\"input/bedFiles_various_genefeatures\"  # Update input folder\n",
    "utr3 = pd.read_csv(f\"{input_folder}/3UTR\", sep=\"\\t\", index_col=None, header=None)\n",
    "print(f\"utr3.shape: {utr3.shape}\")\n",
    "utr3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of different gene IDs: 22127\n"
     ]
    }
   ],
   "source": [
    "# Unique identifiers\n",
    "unique_ids = pd.unique(utr3[3])\n",
    "print(f\"Number of different gene IDs: {len(unique_ids)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "invalid literal for int() with base 10: '/bin/bash: línea 1: ./BiologicalInterpretation/step2.sh: Permiso denegado'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "Input \u001b[0;32mIn [25]\u001b[0m, in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      4\u001b[0m input_file \u001b[38;5;241m=\u001b[39m \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;132;01m{\u001b[39;00minput_folder\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m/\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mevaluate\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m      5\u001b[0m n_segments \u001b[38;5;241m=\u001b[39m get_ipython()\u001b[38;5;241m.\u001b[39mgetoutput(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m./BiologicalInterpretation/step2.sh $min_overlap $input_file\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[0;32m----> 6\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mPerc. of overlapping between working segments and \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mevaluate\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m regions: \u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mround\u001b[39m((\u001b[38;5;28mint\u001b[39m(n_segments[\u001b[38;5;241m0\u001b[39m])\u001b[38;5;241m/\u001b[39m\u001b[38;5;241m8022\u001b[39m)\u001b[38;5;241m*\u001b[39m\u001b[38;5;241m100\u001b[39m, \u001b[38;5;241m2\u001b[39m)\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m%\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "\u001b[0;31mValueError\u001b[0m: invalid literal for int() with base 10: '/bin/bash: línea 1: ./BiologicalInterpretation/step2.sh: Permiso denegado'"
     ]
    }
   ],
   "source": [
    "min_overlap=0.03\n",
    "input_folder = \"bedFiles_various_genefeatures\"\n",
    "evaluate = \"3UTR\"\n",
    "input_file = f\"{input_folder}/{evaluate}\"\n",
    "n_segments = !./BiologicalInterpretation/step2.sh $min_overlap $input_file\n",
    "print(f\"Perc. of overlapping between working segments and {evaluate} regions: {round((int(n_segments[0])/8022)*100, 2)}%\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### **5UTR**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "utr5.shape: (31098, 6)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>66999835</td>\n",
       "      <td>67000041</td>\n",
       "      <td>NM_032291.4_utr5_0_0_chr1_66999836_f</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>8378173</td>\n",
       "      <td>8378246</td>\n",
       "      <td>NM_001080397.3_utr5_0_0_chr1_8378174_f</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>8384365</td>\n",
       "      <td>8384389</td>\n",
       "      <td>NM_001080397.3_utr5_1_0_chr1_8384366_f</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>16767218</td>\n",
       "      <td>16767256</td>\n",
       "      <td>NM_018090.5_utr5_0_0_chr1_16767219_f</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>25071937</td>\n",
       "      <td>25072044</td>\n",
       "      <td>NM_013943.3_utr5_0_0_chr1_25071938_f</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31093</th>\n",
       "      <td>chr22_kb663609_alt</td>\n",
       "      <td>2044</td>\n",
       "      <td>2113</td>\n",
       "      <td>NM_033318.5_utr5_0_0_chr22_kb663609_alt_2045_f</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31094</th>\n",
       "      <td>chr22_kb663609_alt</td>\n",
       "      <td>13089</td>\n",
       "      <td>13117</td>\n",
       "      <td>NM_002490.6_utr5_2_0_chr22_kb663609_alt_13090_r</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31095</th>\n",
       "      <td>chr22_kb663609_alt</td>\n",
       "      <td>53118</td>\n",
       "      <td>53137</td>\n",
       "      <td>NM_000106.6_utr5_8_0_chr22_kb663609_alt_53119_r</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31096</th>\n",
       "      <td>chr7_gl000195_random</td>\n",
       "      <td>49117</td>\n",
       "      <td>49119</td>\n",
       "      <td>NM_001347681.2_utr5_1_0_chr7_gl000195_random_4...</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31097</th>\n",
       "      <td>chr7_gl000195_random</td>\n",
       "      <td>86517</td>\n",
       "      <td>86735</td>\n",
       "      <td>NM_001347681.2_utr5_2_0_chr7_gl000195_random_8...</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>31098 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          0         1         2  \\\n",
       "0                      chr1  66999835  67000041   \n",
       "1                      chr1   8378173   8378246   \n",
       "2                      chr1   8384365   8384389   \n",
       "3                      chr1  16767218  16767256   \n",
       "4                      chr1  25071937  25072044   \n",
       "...                     ...       ...       ...   \n",
       "31093    chr22_kb663609_alt      2044      2113   \n",
       "31094    chr22_kb663609_alt     13089     13117   \n",
       "31095    chr22_kb663609_alt     53118     53137   \n",
       "31096  chr7_gl000195_random     49117     49119   \n",
       "31097  chr7_gl000195_random     86517     86735   \n",
       "\n",
       "                                                       3  4  5  \n",
       "0                   NM_032291.4_utr5_0_0_chr1_66999836_f  0  +  \n",
       "1                 NM_001080397.3_utr5_0_0_chr1_8378174_f  0  +  \n",
       "2                 NM_001080397.3_utr5_1_0_chr1_8384366_f  0  +  \n",
       "3                   NM_018090.5_utr5_0_0_chr1_16767219_f  0  +  \n",
       "4                   NM_013943.3_utr5_0_0_chr1_25071938_f  0  +  \n",
       "...                                                  ... .. ..  \n",
       "31093     NM_033318.5_utr5_0_0_chr22_kb663609_alt_2045_f  0  +  \n",
       "31094    NM_002490.6_utr5_2_0_chr22_kb663609_alt_13090_r  0  -  \n",
       "31095    NM_000106.6_utr5_8_0_chr22_kb663609_alt_53119_r  0  -  \n",
       "31096  NM_001347681.2_utr5_1_0_chr7_gl000195_random_4...  0  -  \n",
       "31097  NM_001347681.2_utr5_2_0_chr7_gl000195_random_8...  0  -  \n",
       "\n",
       "[31098 rows x 6 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "input_folder=\"input/bedFiles_various_genefeatures\"  # Update input folder\n",
    "utr5 = pd.read_csv(f\"{input_folder}/5UTR\", sep=\"\\t\", index_col=None, header=None)\n",
    "print(f\"utr5.shape: {utr5.shape}\")\n",
    "utr5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of different gene IDs: 31098\n"
     ]
    }
   ],
   "source": [
    "# Unique identifiers\n",
    "unique_ids = pd.unique(utr5[3])\n",
    "print(f\"Number of different gene IDs: {len(unique_ids)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "invalid literal for int() with base 10: '/bin/bash: línea 1: ./BiologicalInterpretation/step2.sh: Permiso denegado'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "Input \u001b[0;32mIn [28]\u001b[0m, in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      4\u001b[0m input_file \u001b[38;5;241m=\u001b[39m \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;132;01m{\u001b[39;00minput_folder\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m/\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mevaluate\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m      5\u001b[0m n_segments \u001b[38;5;241m=\u001b[39m get_ipython()\u001b[38;5;241m.\u001b[39mgetoutput(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m./BiologicalInterpretation/step2.sh $min_overlap $input_file\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[0;32m----> 6\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mPerc. of overlapping between working segments and \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mevaluate\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m regions: \u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mround\u001b[39m((\u001b[38;5;28mint\u001b[39m(n_segments[\u001b[38;5;241m0\u001b[39m])\u001b[38;5;241m/\u001b[39m\u001b[38;5;241m8022\u001b[39m)\u001b[38;5;241m*\u001b[39m\u001b[38;5;241m100\u001b[39m, \u001b[38;5;241m2\u001b[39m)\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m%\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m      7\u001b[0m utr3\n",
      "\u001b[0;31mValueError\u001b[0m: invalid literal for int() with base 10: '/bin/bash: línea 1: ./BiologicalInterpretation/step2.sh: Permiso denegado'"
     ]
    }
   ],
   "source": [
    "# min_overlap=0.001\n",
    "input_folder = \"bedFiles_various_genefeatures\"\n",
    "evaluate = \"5UTR\"\n",
    "input_file = f\"{input_folder}/{evaluate}\"\n",
    "n_segments = !./BiologicalInterpretation/step2.sh $min_overlap $input_file\n",
    "print(f\"Perc. of overlapping between working segments and {evaluate} regions: {round((int(n_segments[0])/8022)*100, 2)}%\")\n",
    "utr3"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### **Introns**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "introns.shape: (200135, 6)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>67000051</td>\n",
       "      <td>67091529</td>\n",
       "      <td>NM_032291.4_intron_0_0_chr1_67000052_f</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>67091593</td>\n",
       "      <td>67098752</td>\n",
       "      <td>NM_032291.4_intron_1_0_chr1_67091594_f</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>67098777</td>\n",
       "      <td>67101626</td>\n",
       "      <td>NM_032291.4_intron_2_0_chr1_67098778_f</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>67101698</td>\n",
       "      <td>67105459</td>\n",
       "      <td>NM_032291.4_intron_3_0_chr1_67101699_f</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>67105516</td>\n",
       "      <td>67108492</td>\n",
       "      <td>NM_032291.4_intron_4_0_chr1_67105517_f</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>200130</th>\n",
       "      <td>chr22_kb663609_alt</td>\n",
       "      <td>51271</td>\n",
       "      <td>51359</td>\n",
       "      <td>NM_000106.6_intron_5_0_chr22_kb663609_alt_51272_r</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>200131</th>\n",
       "      <td>chr22_kb663609_alt</td>\n",
       "      <td>51512</td>\n",
       "      <td>52064</td>\n",
       "      <td>NM_000106.6_intron_6_0_chr22_kb663609_alt_51513_r</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>200132</th>\n",
       "      <td>chr22_kb663609_alt</td>\n",
       "      <td>52236</td>\n",
       "      <td>52938</td>\n",
       "      <td>NM_000106.6_intron_7_0_chr22_kb663609_alt_52237_r</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>200133</th>\n",
       "      <td>chr7_gl000195_random</td>\n",
       "      <td>44923</td>\n",
       "      <td>48954</td>\n",
       "      <td>NM_001347681.2_intron_0_0_chr7_gl000195_random...</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>200134</th>\n",
       "      <td>chr7_gl000195_random</td>\n",
       "      <td>49119</td>\n",
       "      <td>86517</td>\n",
       "      <td>NM_001347681.2_intron_1_0_chr7_gl000195_random...</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>200135 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           0         1         2  \\\n",
       "0                       chr1  67000051  67091529   \n",
       "1                       chr1  67091593  67098752   \n",
       "2                       chr1  67098777  67101626   \n",
       "3                       chr1  67101698  67105459   \n",
       "4                       chr1  67105516  67108492   \n",
       "...                      ...       ...       ...   \n",
       "200130    chr22_kb663609_alt     51271     51359   \n",
       "200131    chr22_kb663609_alt     51512     52064   \n",
       "200132    chr22_kb663609_alt     52236     52938   \n",
       "200133  chr7_gl000195_random     44923     48954   \n",
       "200134  chr7_gl000195_random     49119     86517   \n",
       "\n",
       "                                                        3  4  5  \n",
       "0                  NM_032291.4_intron_0_0_chr1_67000052_f  0  +  \n",
       "1                  NM_032291.4_intron_1_0_chr1_67091594_f  0  +  \n",
       "2                  NM_032291.4_intron_2_0_chr1_67098778_f  0  +  \n",
       "3                  NM_032291.4_intron_3_0_chr1_67101699_f  0  +  \n",
       "4                  NM_032291.4_intron_4_0_chr1_67105517_f  0  +  \n",
       "...                                                   ... .. ..  \n",
       "200130  NM_000106.6_intron_5_0_chr22_kb663609_alt_51272_r  0  -  \n",
       "200131  NM_000106.6_intron_6_0_chr22_kb663609_alt_51513_r  0  -  \n",
       "200132  NM_000106.6_intron_7_0_chr22_kb663609_alt_52237_r  0  -  \n",
       "200133  NM_001347681.2_intron_0_0_chr7_gl000195_random...  0  -  \n",
       "200134  NM_001347681.2_intron_1_0_chr7_gl000195_random...  0  -  \n",
       "\n",
       "[200135 rows x 6 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "input_folder=\"input/bedFiles_various_genefeatures\"  # Update input folder\n",
    "introns = pd.read_csv(f\"{input_folder}/introns\", sep=\"\\t\", index_col=None, header=None)\n",
    "print(f\"introns.shape: {introns.shape}\")\n",
    "introns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "invalid literal for int() with base 10: '/bin/bash: línea 1: ./BiologicalInterpretation/step2.sh: Permiso denegado'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "Input \u001b[0;32mIn [30]\u001b[0m, in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      4\u001b[0m input_file \u001b[38;5;241m=\u001b[39m \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;132;01m{\u001b[39;00minput_folder\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m/\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mevaluate\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m      5\u001b[0m n_segments \u001b[38;5;241m=\u001b[39m get_ipython()\u001b[38;5;241m.\u001b[39mgetoutput(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m./BiologicalInterpretation/step2.sh $min_overlap $input_file\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[0;32m----> 6\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mPerc. of overlapping between working segments and \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mevaluate\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m regions: \u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mround\u001b[39m((\u001b[38;5;28mint\u001b[39m(n_segments[\u001b[38;5;241m0\u001b[39m])\u001b[38;5;241m/\u001b[39m\u001b[38;5;241m8022\u001b[39m)\u001b[38;5;241m*\u001b[39m\u001b[38;5;241m100\u001b[39m, \u001b[38;5;241m2\u001b[39m)\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m%\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "\u001b[0;31mValueError\u001b[0m: invalid literal for int() with base 10: '/bin/bash: línea 1: ./BiologicalInterpretation/step2.sh: Permiso denegado'"
     ]
    }
   ],
   "source": [
    "# min_overlap=0.001\n",
    "input_folder = \"bedFiles_various_genefeatures\"\n",
    "evaluate = \"introns\"\n",
    "input_file = f\"{input_folder}/{evaluate}\"\n",
    "n_segments = !./BiologicalInterpretation/step2.sh $min_overlap $input_file\n",
    "print(f\"Perc. of overlapping between working segments and {evaluate} regions: {round((int(n_segments[0])/8022)*100, 2)}%\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of different gene IDs: 200135\n"
     ]
    }
   ],
   "source": [
    "# Unique identifiers\n",
    "unique_ids = pd.unique(introns[3])\n",
    "print(f\"Number of different gene IDs: {len(unique_ids)}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### **Upstream region of genes**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "upstreamGenes.shape: (21458, 6)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>66999635</td>\n",
       "      <td>66999835</td>\n",
       "      <td>NM_032291.4_up_200_chr1_66999636_f</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>8377973</td>\n",
       "      <td>8378173</td>\n",
       "      <td>NM_001080397.3_up_200_chr1_8377974_f</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>16767018</td>\n",
       "      <td>16767218</td>\n",
       "      <td>NM_018090.5_up_200_chr1_16767019_f</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>25071737</td>\n",
       "      <td>25071937</td>\n",
       "      <td>NM_013943.3_up_200_chr1_25071738_f</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>33546553</td>\n",
       "      <td>33546753</td>\n",
       "      <td>NM_052998.4_up_200_chr1_33546554_f</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21453</th>\n",
       "      <td>chr22_kb663609_alt</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>NM_001002034.3_up_200_chr22_kb663609_alt_-198_f</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21454</th>\n",
       "      <td>chr22_kb663609_alt</td>\n",
       "      <td>1844</td>\n",
       "      <td>2044</td>\n",
       "      <td>NM_033318.5_up_200_chr22_kb663609_alt_1845_f</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21455</th>\n",
       "      <td>chr22_kb663609_alt</td>\n",
       "      <td>13117</td>\n",
       "      <td>13317</td>\n",
       "      <td>NM_002490.6_up_200_chr22_kb663609_alt_13118_r</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21456</th>\n",
       "      <td>chr22_kb663609_alt</td>\n",
       "      <td>53137</td>\n",
       "      <td>53337</td>\n",
       "      <td>NM_000106.6_up_200_chr22_kb663609_alt_53138_r</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21457</th>\n",
       "      <td>chr7_gl000195_random</td>\n",
       "      <td>86735</td>\n",
       "      <td>86935</td>\n",
       "      <td>NM_001347681.2_up_200_chr7_gl000195_random_867...</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>21458 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          0         1         2  \\\n",
       "0                      chr1  66999635  66999835   \n",
       "1                      chr1   8377973   8378173   \n",
       "2                      chr1  16767018  16767218   \n",
       "3                      chr1  25071737  25071937   \n",
       "4                      chr1  33546553  33546753   \n",
       "...                     ...       ...       ...   \n",
       "21453    chr22_kb663609_alt         0         1   \n",
       "21454    chr22_kb663609_alt      1844      2044   \n",
       "21455    chr22_kb663609_alt     13117     13317   \n",
       "21456    chr22_kb663609_alt     53137     53337   \n",
       "21457  chr7_gl000195_random     86735     86935   \n",
       "\n",
       "                                                       3  4  5  \n",
       "0                     NM_032291.4_up_200_chr1_66999636_f  0  +  \n",
       "1                   NM_001080397.3_up_200_chr1_8377974_f  0  +  \n",
       "2                     NM_018090.5_up_200_chr1_16767019_f  0  +  \n",
       "3                     NM_013943.3_up_200_chr1_25071738_f  0  +  \n",
       "4                     NM_052998.4_up_200_chr1_33546554_f  0  +  \n",
       "...                                                  ... .. ..  \n",
       "21453    NM_001002034.3_up_200_chr22_kb663609_alt_-198_f  0  +  \n",
       "21454       NM_033318.5_up_200_chr22_kb663609_alt_1845_f  0  +  \n",
       "21455      NM_002490.6_up_200_chr22_kb663609_alt_13118_r  0  -  \n",
       "21456      NM_000106.6_up_200_chr22_kb663609_alt_53138_r  0  -  \n",
       "21457  NM_001347681.2_up_200_chr7_gl000195_random_867...  0  -  \n",
       "\n",
       "[21458 rows x 6 columns]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "input_folder=\"input/bedFiles_various_genefeatures\"  # Update input folder\n",
    "upstreamGenes = pd.read_csv(f\"{input_folder}/upstreamGenes\", sep=\"\\t\", index_col=None, header=None)\n",
    "print(f\"upstreamGenes.shape: {upstreamGenes.shape}\")\n",
    "upstreamGenes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of different gene IDs: 21458\n"
     ]
    }
   ],
   "source": [
    "# Unique identifiers\n",
    "unique_ids = pd.unique(upstreamGenes[3])\n",
    "print(f\"Number of different gene IDs: {len(unique_ids)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "invalid literal for int() with base 10: '/bin/bash: línea 1: ./BiologicalInterpretation/step2.sh: Permiso denegado'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "Input \u001b[0;32mIn [34]\u001b[0m, in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      4\u001b[0m input_file \u001b[38;5;241m=\u001b[39m \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;132;01m{\u001b[39;00minput_folder\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m/\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mevaluate\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m      5\u001b[0m n_segments \u001b[38;5;241m=\u001b[39m get_ipython()\u001b[38;5;241m.\u001b[39mgetoutput(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m./BiologicalInterpretation/step2.sh $min_overlap $input_file\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[0;32m----> 6\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mPerc. of overlapping between working segments and \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mevaluate\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m regions: \u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mround\u001b[39m((\u001b[38;5;28mint\u001b[39m(n_segments[\u001b[38;5;241m0\u001b[39m])\u001b[38;5;241m/\u001b[39m\u001b[38;5;241m8022\u001b[39m)\u001b[38;5;241m*\u001b[39m\u001b[38;5;241m100\u001b[39m, \u001b[38;5;241m2\u001b[39m)\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m%\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "\u001b[0;31mValueError\u001b[0m: invalid literal for int() with base 10: '/bin/bash: línea 1: ./BiologicalInterpretation/step2.sh: Permiso denegado'"
     ]
    }
   ],
   "source": [
    "# min_overlap=0.001\n",
    "input_folder = \"bedFiles_various_genefeatures\"\n",
    "evaluate = \"upstreamGenes\"\n",
    "input_file = f\"{input_folder}/{evaluate}\"\n",
    "n_segments = !./BiologicalInterpretation/step2.sh $min_overlap $input_file\n",
    "print(f\"Perc. of overlapping between working segments and {evaluate} regions: {round((int(n_segments[0])/8022)*100, 2)}%\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### (3) **Methylation**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "invalid literal for int() with base 10: '/bin/bash: línea 1: ./BiologicalInterpretation/step2.sh: Permiso denegado'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "Input \u001b[0;32mIn [35]\u001b[0m, in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      4\u001b[0m input_file \u001b[38;5;241m=\u001b[39m \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;132;01m{\u001b[39;00minput_folder\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m/\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mevaluate\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m      5\u001b[0m n_segments \u001b[38;5;241m=\u001b[39m get_ipython()\u001b[38;5;241m.\u001b[39mgetoutput(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m./BiologicalInterpretation/step2.sh $min_overlap $input_file\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[0;32m----> 6\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mPerc. of overlapping between working segments and \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mevaluate\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m regions: \u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mround\u001b[39m((\u001b[38;5;28mint\u001b[39m(n_segments[\u001b[38;5;241m0\u001b[39m])\u001b[38;5;241m/\u001b[39m\u001b[38;5;241m8022\u001b[39m)\u001b[38;5;241m*\u001b[39m\u001b[38;5;241m100\u001b[39m, \u001b[38;5;241m2\u001b[39m)\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m%\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "\u001b[0;31mValueError\u001b[0m: invalid literal for int() with base 10: '/bin/bash: línea 1: ./BiologicalInterpretation/step2.sh: Permiso denegado'"
     ]
    }
   ],
   "source": [
    "# min_overlap=0.001\n",
    "input_folder = \"forExercise6\"\n",
    "evaluate = \"hypoBED\"\n",
    "input_file = f\"{input_folder}/{evaluate}\"\n",
    "n_segments = !./BiologicalInterpretation/step2.sh $min_overlap $input_file\n",
    "print(f\"Perc. of overlapping between working segments and {evaluate} regions: {round((int(n_segments[0])/8022)*100, 2)}%\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "invalid literal for int() with base 10: '/bin/bash: línea 1: ./BiologicalInterpretation/step2.sh: Permiso denegado'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "Input \u001b[0;32mIn [36]\u001b[0m, in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      4\u001b[0m input_file \u001b[38;5;241m=\u001b[39m \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;132;01m{\u001b[39;00minput_folder\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m/\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mevaluate\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m      5\u001b[0m n_segments \u001b[38;5;241m=\u001b[39m get_ipython()\u001b[38;5;241m.\u001b[39mgetoutput(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m./BiologicalInterpretation/step2.sh $min_overlap $input_file\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[0;32m----> 6\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mPerc. of overlapping between working segments and \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mevaluate\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m regions: \u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mround\u001b[39m((\u001b[38;5;28mint\u001b[39m(n_segments[\u001b[38;5;241m0\u001b[39m])\u001b[38;5;241m/\u001b[39m\u001b[38;5;241m8022\u001b[39m)\u001b[38;5;241m*\u001b[39m\u001b[38;5;241m100\u001b[39m, \u001b[38;5;241m2\u001b[39m)\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m%\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "\u001b[0;31mValueError\u001b[0m: invalid literal for int() with base 10: '/bin/bash: línea 1: ./BiologicalInterpretation/step2.sh: Permiso denegado'"
     ]
    }
   ],
   "source": [
    "# min_overlap=0.001\n",
    "input_folder = \"forExercise6\"\n",
    "evaluate = \"hyperBED\"\n",
    "input_file = f\"{input_folder}/{evaluate}\"\n",
    "n_segments = !./BiologicalInterpretation/step2.sh $min_overlap $input_file\n",
    "print(f\"Perc. of overlapping between working segments and {evaluate} regions: {round((int(n_segments[0])/8022)*100, 2)}%\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **Summary Table on annotations**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "| **Feature**                                                      | Overlap in terms of length | Overlap in terms of number of segments (no min. overlap) | Overlap in terms of number of segments (-f 0.03) |\n",
    "|------------------------------------------------------------------|----------------------------|----------------------------------------------------------|--------------------------------------------------|\n",
    "| Percentage of coding genes with CTCF segments                    | 39.73%                     |                                                          |                                                  |\n",
    "| Percentage of exons of coding genes with CTCF segments           | 1.52%                      |                                                          |                                                  |\n",
    "| Percentage of introns of coding genes with CTCF segments         | 35.70%                     | 42.2%                                                    | 41.85%                                           |\n",
    "| Percentage of 3 UTR of coding genes with CTCF segments           | 2.36%                      | 3.63%                                                    | 3.48%                                            |\n",
    "| Percentage of 5 UTR of coding genes with CTCF segments           | 0.14%                      | 0.82%                                                    | 0.72%                                            |\n",
    "| Percentage of upstream 200 bp of coding genes with CTCF segments | 0.22%                      | 1.07%                                                    | 1.02%                                            |\n",
    "| Percentage of overlapping with DNaseI                            | 25.88%                     | 55.51%                                                   | 55.19%                                           |\n",
    "| Percentage of overlapping with <span style=\"color:red\">hypermetylated</span> regions           | 80.87%                     | 93.63%                                                   | 92.47%                                           |\n",
    "| Percentage of overlapping with <span style=\"color:red\">hypomethylated</span> regions            | 2.55%                      | 7.21%                                                    | 6.49%                                            |"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Considerations**:\n",
    "\n",
    "* Overlap in terms of length from Inés, overlap in terms of number of segments from Iván (in case of doubts with the approaches).\n",
    "\n",
    "* The approaches are different even if the final percentages are similar: it might not be the same an overlap between one of our working segments and any of the reported features *in an insignificant region of our segment* vs. *in the actual CTCF DNA-binding motif of our segment*. However, we cannot be sure whether we are capturing this CTCF DNA binding motif with any of both approaches.\n",
    "\n",
    "* Hypo/ hypermethylated percentages have to be corrected."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## **References**\n",
    "\n",
    "(0) CTCF summary\n",
    "\n",
    "(1) Lu Y, Shan G, Xue J, Chen C, Zhang C. *Defining the multivalent functions of CTCF from chromatin state and three-dimensional chromatin interactions*. Nucleic Acids Res. 2016 Jul 27;44(13):6200-12. doi: 10.1093/nar/gkw249. Epub 2016 Apr 11. PMID: 27067545; PMCID: PMC5291250. \\label{1}\n",
    "\n",
    "(2) Xi H, Shulha HP, Lin JM, Vales TR, Fu Y, Bodine DM, et al. (2007) *Identification and Characterization of Cell Type–Specific and Ubiquitous Chromatin Regulatory Structures in the Human Genome*. PLoS Genet 3(8): e136. https://doi.org/10.1371/journal.pgen.0030136. \\label{2}\n",
    "\n",
    "(3) Cuddapah S, Jothi R, Schones DE, Roh TY, Cui K, Zhao K. *Global analysis of the insulator binding protein CTCF in chromatin barrier regions reveals demarcation of active and repressive domains*. Genome Res. 2009 Jan;19(1):24-32. doi: 10.1101/gr.082800.108. Epub 2008 Dec 3. PMID: 19056695; PMCID: PMC2612964. \\label{3}\n",
    "\n",
    "(4) Ong, CT., Corces, V. *CTCF: an architectural protein bridging genome topology and function*. Nat Rev Genet 15, 234–246 (2014). https://doi.org/10.1038/nrg3663. \\label{4}\n",
    "\n",
    "(5) Ghirlando R, Felsenfeld G. CTCF: making the right connections. Genes Dev. 2016 Apr 15;30(8):881-91. doi: 10.1101/gad.277863.116. PMID: 27083996; PMCID: PMC4840295. \\label{5}"
   ]
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "e4ee870ab444af8a8689fba9fdb6a16993f9af4d6f8c51486b98fd7ee4129479"
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
