# CTCFepigenomics
Epigenomic analysis of CTCF marker, which is associated to insulators.

1) Obtain the segments that have the same state in the two replicates of monocytes. Cehck: *'updated_analysis.ipynb'*.
2) Annotate the segments. As a minimum, the percentage of segments that overlap with protein-coding genes. Check *'DNAse1_peak_overlapping.docx'*.
3) Download the DNase I peaks for monocytes from ENCODE and calculate the percentage of overlap between DNaseI-peaks and your working segments. Use the wgEncodeOpenChromDnaseMonocd14Pk.narrowPeak.gz file in: http://hgdownload.cse.ucsc.edu/goldenpath/hg19/encodeDCC/wgEncodeOpenChromDnase/ See *'summaryBiologicalInterpretation.ipynb'*. 
4) Visualize (and show) a region of the genome in the UCSC browser where you see at least one of your segments (upload the track generated by ChromHMM) and the DNaseI in all ENCODE blood cell types. CD34+ cells (enriched in hematopoietic stem and progenitor cells). Check *'DNAse1_peak_overlapping.docx'*.
5) Search for enriched motifs. See *'summaryBiologicalInterpretation.ipynb'*.
6) Calculate the % of overlap with hyper- (Methylation> 0.75) or hypo-methylated (Methylation <0.25) regions in monocytes available in the BLUEPRINT portal DCC. (http://dcc.blueprint-epigenome.eu/#/home) BED files belonging to donor C001UY.See *'summaryBiologicalInterpretation.ipynb'*.
